### PR TITLE
[move] Rewrite verifier metering

### PR DESCRIPTION
**Pull request by tnowacki**
## Description 

- Added ability cache for verifier ability queries
- Redid metering for typing, local safety, and reference safety 

## Test plan 

- calibrated constants

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:
- [ ] REST API:

### CODE REVIEW

File: external-crates/move/crates/move-bytecode-verifier/src/reference_safety/abstract_state.rs [hunk 1]
```diff
@@ -166,27 +164,63 @@ impl AbstractState {
...
+        &mut self,
+        resource: StructDefinitionIndex,
+        id: RefID,
+        meter: &mut (impl Meter + ?Sized),
+    ) -> PartialVMResult<()> {
+        meter.add(Scope::Function, ADD_BORROW_COST)?;
         self.borrow_graph
-            .add_weak_field_borrow((), self.frame_root(), Label::Global(resource), id)
+            .add_weak_field_borrow((), self.frame_root(), Label::Global(resource), id);
+        Ok(())
```
Comment (tzakian) [inline external-crates/move/crates/move-bytecode-verifier/src/reference_safety/abstract_state.rs:223]: Should add charging to the `add_variant_field_borrow` below as well I believe.

File: external-crates/move/crates/move-bytecode-verifier/src/reference_safety/abstract_state.rs [hunk 2]
```diff
@@ -801,3 +907,40 @@ impl AbstractDomain for AbstractState {
         }
     }
 }
+
+fn charge_graph_size(size: usize, meter: &mut (impl Meter + ?Sized)) -> PartialVMResult<()> {
+    let size = max(size, 1);
```
Comment (tzakian) [inline external-crates/move/crates/move-bytecode-verifier/src/reference_safety/abstract_state.rs:912]: The smallest of nits but some places you use `std::cmp::max` and here you use `max`. Don't have a preference for either one, but probably best to be consistent.

File: external-crates/move/crates/move-bytecode-verifier/src/signature.rs [hunk 3]
```diff
@@ -15,22 +15,41 @@ use move_binary_format::{
...
+};
+
+use crate::ability_cache::AbilityCache;
 
-pub struct SignatureChecker<'a> {
-    module: &'a CompiledModule,
+pub struct SignatureChecker<'env, 'a, 'b, M: Meter + ?Sized> {
+    module: &'env CompiledModule,
+    module_ability_cache: RefCell<&'a mut AbilityCache<'env>>,
+    meter: RefCell<&'b mut M>,
```
Comment (tzakian) [inline external-crates/move/crates/move-bytecode-verifier/src/signature.rs:30]: Is there a particular reason for these `RefCell`s? Would be nice if they weren't needed.

**Review by tzakian — COMMENTED** [review]
Overall looks good to me, just a couple (mainly smaller) comments.

File: external-crates/move/crates/move-binary-format/src/lib.rs [hunk 4]
```diff
@@ -161,8 +161,10 @@ macro_rules! safe_unwrap {
         match $e {
             Some(x) => x,
             None => {
-                let err = PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
-                    .with_message(format!("{}:{} (none)", file!(), line!()));
+                let err = move_binary_format::errors::PartialVMError::new(
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply external-crates/move/crates/move-binary-format/src/lib.rs:164]: This is in a macro declaration. I feel like maybe the fully qualified paths are more clear?

File: external-crates/move/crates/move-bytecode-verifier/src/signature.rs (hunk 3 above)
Comment (tnowacki) [reply external-crates/move/crates/move-bytecode-verifier/src/signature.rs]: laziness, fixed

File: external-crates/move/crates/move-bytecode-verifier/src/locals_safety/abstract_state.rs [hunk 5]
```diff
@@ -26,10 +26,11 @@ pub(crate) enum LocalState {
 }
 use LocalState::*;
 
-pub(crate) const STEP_BASE_COST: u128 = 15;
-pub(crate) const RET_PER_LOCAL_COST: u128 = 30;
-pub(crate) const JOIN_BASE_COST: u128 = 10;
-pub(crate) const JOIN_PER_LOCAL_COST: u128 = 5;
+use crate::ability_cache::AbilityCache;
```
[the remark this answers is not in the document]
Comment (tnowacki) [reply external-crates/move/crates/move-bytecode-verifier/src/locals_safety/abstract_state.rs:29]: Remember what we were discussing about autoimports in the IDE... yeah this is why its hard

File: external-crates/move/crates/move-binary-format/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tzakian) [reply external-crates/move/crates/move-binary-format/src/lib.rs:164]: ah! didn't realize that. Makes sense to keep as-is then.

**Review by tzakian — APPROVED** [review]
Looks great!

**Pull request merged**